Action to run accessibility tests and create markdown report (#25)
Quansight-Labs/jupyter-a11y-testing: testing/jupyterlab/test-info-pages/no-tab-trap-initial-page.md

> # No Tab Trap on Initial Page

Author: My vision here is to a have a Markdown "info page" for every regression test, explaining what fix it checks, why that fix is important, what accessibility criteria it relates to, how to perform the test manually, etc. 

I want something like the [ACT Rule Template](https://act-rules.github.io/pages/design/rule-template/) but I want the format to be a bit more lightweight, if possible, so that it's not too burdensome to add a new regression test. I'm thinking that the final format will become evident after we've added about 10 new tests to the repo.

> # No Tab Trap on Initial Page

Author: I wasn't sure what to call these documents. I didn't really like the name "testing scripts" because "scripts" usually means a code file that automates a task, which is the opposite of how we mean it here.

Some alternatives I considered:

- test docs, which I didn't like because I think people might think that they are the docs for the repo or for the whole testing apparatus
- test wiki pages - no because I don't think these are really a wiki, the scope is too narrow
- test descriptions - this is a pretty accurate name but I worry that it's not specific enough because there are other things in the system that somebody might refer to as "test descriptions" for example the text in `test.describe` blocks. I want these things to be called something that when people say the word, they know you're talking about this and not that.
- test summaries - but then I think people will think these are about the results of test output
- test resumes - I kind of liked this one but I also worried if the word resume is too close to summary. Like, again, I don't want folks thinking that this is about summarizing the outputs of the test runs.
- test profiles - but then devs might think we're talking about code performance profiling (CPU, memory usage)

Reviewer: I don't have a strong opinion, but I appreciate that you documented possible confusion points. Info and/or info pages works for me.

Reviewer 2: these seem to fall under manual testing scripts (which is the official and industry-wide adopted term)
it is often better to look at what the standard is than creating a new pattern and accounting for all possible confusion from folks

Author: Oh interesting! I didn't know that. @trallard could you point me to a variety of bodies of manual testing scripts? That way I can check if the format of those is close enough to the format of what I'm proposing

Reviewer 2: I searched for some good examples with no avail - since most of these tests I have seen are internal documents, I have no real-world examples to share. But generally, they have sections like

- Metadata: author, date updated, version, etc.
- Description: what this is, what does it test, and any compliance/requirements needed
- A table with info like
  - Steps (instructions)
  - Inputs (if any)
  - Expected outcome

> - id: 1wlgh6

Reviewer 2: where is this id created from?

Author: [random.org/strings](https://www.random.org/strings/?num=1&len=6&digits=on&loweralpha=on&format=html&rnd=new)

ACT [uses random ids](https://act-rules.github.io/pages/design/rule-design/) for their rules. That way the name of the test can change but it keeps the same id.

## Description

When a user loads the app for the first time, they should be able to use the tab

> key to navigate without getting trapped.

Reviewer: ```suggestion
key to navigate the page they land on without getting trapped.
```

## Applicability

Which apps does this test apply to.

- JupyterLab

## Accessibility Criteria

This test is related to:

- [WCAG Success Criterion 2.1.2 No Keyboard
  Trap](https://www.w3.org/TR/WCAG22/#no-keyboard-trap)

If this test fails, it means the app fails to conform to 2.1.2.

> If this test passes, however, it does not mean that the app conforms to 2.1.2.

Author: I think most of the regression tests will be like this, i.e., they can tell you if you've violated a WCAG criterion but not if the entire app conforms to the criterion, so we might not need to explain this in every test info page. We might need to have one page of definitions and explanations that each individual test info page links back to.

Reviewer 2: I find the whole way this is phrased a tad confusing:

- Fail -> no conformance
- Pass -> still no conformance? what does this mean concerning the criterion? what can I verify then?

Author: Yeah that's the thing about many of the automated tests: they can only tell you if you fail to conform to WCAG.

But I'll reword this section to be less abstract.

It only means that the app does not violate 2.1.2 for this specific test. This

> test can be combined with others to increase confidence that the app as a whole

Reviewer: ```suggestion
test can be combined with tab trap tests for other pages or page states to increase confidence that the app as a whole
```

If the intent is to have this as a general disclaimer, then I think what you have works. If you want to give an idea on this test specifically, I think it would be nice to get a hint of why a passing test does not equate to conformance and/or what would be more thorough.

I'm trying to think of this as if I'm someone without a lot of past web accessibility experience, and I could imagine myself being like "well, what does the test do if it doesn't tell me if I am complying with the standard?" So I imagine pointing out the "why" (like how this only tests one page of a whole app) could help.

Author: Oh yes, I like this suggestion. We can have some general abstract language somewhere else but in the test info page we should make it as concrete as possible.

Reviewer 2: aha Isabela added on my thoughts!

What I want to know is: what does this test tell me? what have I learned?

Author: 👍 love this, will reword

conforms to 2.1.2.

> ## How to Verify the Test Manually

Reviewer: I really like this section! Out of curiosity, are you basing this on any manual testing resource, or is this just a reflection of how you learned to test it?

Author: I'm not quite sure I understand what you're asking. Are you asking specifically about the technique I'm using to test for no tab traps (i.e., cycle back to the beginning) or are you asking about the way I wrote it up?

In either case, it's all my own work. I don't recall getting any of this from anywhere else.

> ## How to Verify the Test Manually

Author: ```suggestion
## How to Perform the Test Manually
```
